lxd/instance.go

```diff
@@ -416,7 +416,73 @@ func instanceLoadNodeProjectAll(ctx context.Context, s *state.State, project str
 	return instances, nil
 }
 
-func autoCreateInstanceSnapshotsTask(d *Daemon) (task.Func, task.Schedule) {
+func autoCreateInstanceSnapshots(ctx context.Context, d *Daemon, instances []instance.Instance) error {
+	// Make the snapshots.
+	for _, inst := range instances {
+		if ctx.Err() != nil {
+			return ctx.Err()
+		}
+
+		ch := make(chan struct{})
+		go func(inst instance.Instance) {
+			defer close(ch)
+			l := logger.AddContext(logger.Log, logger.Ctx{"project": inst.Project().Name, "instance": inst.Name()})
+
+			snapshotName, err := instance.NextSnapshotName(d.State(), inst, "snap%d")
+			if err != nil {
+				l.Error("Error retrieving next snapshot name", logger.Ctx{"err": err})
+				ch <- struct{}{}
+				return
+			}
+
+			expiry, err := shared.GetExpiry(time.Now(), inst.ExpandedConfig()["snapshots.expiry"])
+			if err != nil {
+				l.Error("Error getting snapshots.expiry date")
+				ch <- struct{}{}
+				return
+			}
+
+			err = inst.Snapshot(snapshotName, expiry, false)
+			if err != nil {
+				l.Error("Error creating snapshot", logger.Ctx{"snapshot": snapshotName, "err": err})
+			}
+
+			ch <- struct{}{}
+		}(inst)
+		select {
+		case <-ctx.Done():
```

Comment:
This bit needs to go otherwise it will still return early when the context is cancelled, just replace with:

```go
<-ch
```

With no need for select at all.

Reply:
Actually we dont need the go routine at all now, just the ctx.Err check

Reply:
@gabrielmougard this part hasn't been addressed yet.